woocommerce/woocommerce-ios
# Use DiffableDataSource on OrderListViewModel and OrderListViewController

Part of the DiffableDataSource fix for #1543. This is a continuation of #2797.

⚠️ Depends on #2797. Please review that first. 🙂    

## What

Continuing the plan mentioned in #2797, this implements DiffableDataSource and `FetchResultSnapshotsProvider` on the new (#2797) `OrderListViewController` and `OrderListViewModel` classes. 

![image](https://user-images.githubusercontent.com/198826/92823166-3648d280-f38a-11ea-8b17-b9e81f6fd4d3.png)

### OrderListViewModel Key Changes

`OrderListViewModel` now declares a `snapshotsProvider` which uses the same conditions as the old [`OrdersViewModel.resultsController`](https://github.com/woocommerce/woocommerce-ios/blob/c720e7415d83df7412237d304d6d8fc3c16c8a35/WooCommerce/Classes/ViewRelated/Orders/OrdersViewModel.swift#L52-L76). 

https://github.com/woocommerce/woocommerce-ios/blob/a13bf7699bcc76576100ebbfc67586603705852d/WooCommerce/Classes/ViewRelated/Orders/OrderListViewModel.swift#L53-L75

It's not accessible by the view controller. The view controller will only care about subscribing to new snapshots which are emitted by:

https://github.com/woocommerce/woocommerce-ios/blob/a13bf7699bcc76576100ebbfc67586603705852d/WooCommerce/Classes/ViewRelated/Orders/OrderListViewModel.swift#L77-L80

### OrderListViewController Key Changes

`OrderListViewController` now declares a `DiffableDataSource` which is bound to `tableView.dataSource`. 

https://github.com/woocommerce/woocommerce-ios/blob/a13bf7699bcc76576100ebbfc67586603705852d/WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift#L38-L46

In order to _feed_ new data to this `DiffableDataSource`, we only need to subscribe to `OrderListViewModel.snapshot`:

https://github.com/woocommerce/woocommerce-ios/blob/a13bf7699bcc76576100ebbfc67586603705852d/WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift#L201-L204

This subscription is how the `UITableView` will update its sections and rows.

## Key Concepts

### The DiffableDataSource Is The Source of Truth

A key difference between `DiffableDataSource` and the `FetchedResultsController` we used is that the `DiffableDataSource` is the source of truth. Once it has a snapshot, `DiffableDataSource` **does not need** the `FetchedResultsController` to determine how it will present its data. 

Question | Old FRC Implementation | New DiffableDataSource |
--------|-------|---
How many sections?        |    FRC ([sections.count](https://github.com/woocommerce/woocommerce-ios/blob/a13bf7699bcc76576100ebbfc67586603705852d/WooCommerce/Classes/ViewRelated/Orders/OrdersViewModel.swift#L209-L213))   | DataSource (automatically handled)
How many rows in a section? | FRC ([sectionInfo.numberOfObjects](https://github.com/woocommerce/woocommerce-ios/blob/a13bf7699bcc76576100ebbfc67586603705852d/WooCommerce/Classes/ViewRelated/Orders/OrdersViewModel.swift#L215-L219)) | DataSource (automatically handled)
What object is at this `indexPath`? | FRC ([objectIndex](https://github.com/woocommerce/woocommerce-ios/blob/a13bf7699bcc76576100ebbfc67586603705852d/WooCommerce/Classes/ViewRelated/Orders/OrdersViewModel.swift#L203-L207)) | [DataSource.itemIdenfier](https://github.com/woocommerce/woocommerce-ios/blob/a13bf7699bcc76576100ebbfc67586603705852d/WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift#L515-L518) but with support from [FetchResultSnapshotsProvider.object(withID:)](https://github.com/woocommerce/woocommerce-ios/blob/a13bf7699bcc76576100ebbfc67586603705852d/WooCommerce/Classes/ViewRelated/Orders/OrderListViewModel.swift#L184)

As you can see, in the old FRC implementation, there are two sources of truth: `UITableView` and `FRC`. The `UITableView` keeps its own list of objects and the FRC also keeps its own list of objects. In the new implementation, there is only one: `DiffableDataSource`. In fact, if the `FetchResultSnapshotsProvider` stops working and stops emitting snapshots, the `DiffableDataSource` can still work. It will be presenting stale data but it will still work. 😄 

### Avoid Using Index Paths

We should avoid using index paths. As much as possible, we should use the `ObjectID`. If you look at the implementation in this PR, we interact with the `DiffableDataSource` first before trying to access more data. For example, in the [`didSelectRowAt` delegate method](https://github.com/woocommerce/woocommerce-ios/blob/a13bf7699bcc76576100ebbfc67586603705852d/WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift#L508), we are only given an `indexPath`. In order to retrieve the actual object (`OrderDetailsViewModel`), we retrieve the `ObjectID` from the `DiffableDataSource` first. And then we _fetch_ the object using the `ObjectID`. 

https://github.com/woocommerce/woocommerce-ios/blob/a13bf7699bcc76576100ebbfc67586603705852d/WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift#L515-L518

Fetching the `ObjectID` [uses the `StorageType`](https://github.com/woocommerce/woocommerce-ios/blob/a13bf7699bcc76576100ebbfc67586603705852d/Yosemite/Yosemite/SnapshotsProvider/FetchResultSnapshotsProvider.swift#L140-L149) directly and doesn't care whether the FRC (SnapshotsProvider) is in an incorrect state or not. Thereby avoiding the #1543 crash. 

What happens if we tried to use index paths for retrieving the object instead? That would mean that we're now back to the _old way_ of [accessing the object from the `FRC` using index paths](https://github.com/woocommerce/woocommerce-ios/blob/a13bf7699bcc76576100ebbfc67586603705852d/WooCommerce/Classes/ViewRelated/Orders/OrdersViewModel.swift#L190). And that means we're back to having **two sources of truth**. So if it happens that the FRC and the UITableView are not in sync, we get #1543 again 💥.

The same is true with sections, we should use section identifiers instead of the section index as much as possible. 

https://github.com/woocommerce/woocommerce-ios/blob/a13bf7699bcc76576100ebbfc67586603705852d/WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift#L549-L553

In summary, **never trust** the data in `FetchedResultsController`. Only trust its snapshots and the `DiffableDataSource`. 😄 

## Intentional Bug

In the previous PRs (like #2777), I mentioned that I exposed `obtainPermanentIDs` to fix a bug that is caused by using the snapshots of `NSFetchedResultsController`. I intentionally did not use that in this PR yet in order for you, dear reviewer, to be aware of the bug. 😄 I will fix it in a separate PR. 

For now, expect to see empty cells like this:

<img src="https://user-images.githubusercontent.com/198826/92823650-c5ee8100-f38a-11ea-80cf-f59c3102aeb5.png" width="240">

## Testing

1. Edit the [`viewControllers(for:)` delegate method in `OrdersMasterViewController`](https://github.com/woocommerce/woocommerce-ios/blob/5631e4e1f86a10ead37430d6d8c60f9885aea1be/WooCommerce/Classes/ViewRelated/Orders/OrdersMasterViewController.swift#L79-L81) and replace the contents with:

    ```swift
        if #available(iOS 13.0, *) {
            return makeViewControllers()
        } else {
            return makeDeprecatedViewControllers()
        }
    ```
2. Run the app on a device/simulator with iOS 13.0. 
3. Navigate to Orders. 
4. Confirm that orders are loaded and displayed.
5. Pull to refresh and scroll down to load subsequent pages. Confirm that new pages are loaded. Expect to see empty cells at this point. Also, **notice** that when loading subsequent pages, the animation is a little funky and you seem to be thrown to a different location in the table view. That's another bug that will be fixed with `obtainPermanentIDs`. 😁 
6. Tap on an order. Confirm that the correct order details page is shown. 

## Submitter Checklist
